### ISSUE TITLE
`kubectl diff` "downloads" the OpenAPI at least once per object

### ISSUE THREAD

**Issue by apelisse** [kubernetes/kubernetes#97940]
# How to reproduce

Create three configmaps:
```yaml
apiVersion: v1
data:
  foo: bar
kind: ConfigMap
metadata:
  name: test1
---
apiVersion: v1
data:
  foo: bar
kind: ConfigMap
metadata:
  name: test2
---
apiVersion: v1
data:
  foo: bar
kind: ConfigMap
metadata:
  name: test3
```

`kubectl diff` them (not created, doesn't matter):
```
I0111 16:40:12.434222   45218 loader.go:375] Config loaded from file:  /Users/apelisse/.kube/config
I0111 16:40:12.640029   45218 round_trippers.go:443] GET https://35.202.162.251/openapi/v2?timeout=32s 200 OK in 205 milliseconds
I0111 16:40:12.828759   45218 round_trippers.go:443] GET https://35.202.162.251/openapi/v2?timeout=32s 200 OK in 55 milliseconds
I0111 16:40:13.066990   45218 round_trippers.go:443] GET https://35.202.162.251/api/v1/namespaces/default/configmaps/test1 404 Not Found in 108 milliseconds
I0111 16:40:13.158357   45218 round_trippers.go:443] POST https://35.202.162.251/api/v1/namespaces/default/configmaps?dryRun=All&fieldManager=kubectl-client-side-apply 201 Created in 90 milliseconds
I0111 16:40:13.218538   45218 round_trippers.go:443] GET https://35.202.162.251/openapi/v2?timeout=32s 200 OK in 57 milliseconds
I0111 16:40:13.438977   45218 round_trippers.go:443] GET https://35.202.162.251/api/v1/namespaces/default/configmaps/test2 404 Not Found in 87 milliseconds
I0111 16:40:13.495392   45218 round_trippers.go:443] POST https://35.202.162.251/api/v1/namespaces/default/configmaps?dryRun=All&fieldManager=kubectl-client-side-apply 201 Created in 55 milliseconds
I0111 16:40:13.553417   45218 round_trippers.go:443] GET https://35.202.162.251/openapi/v2?timeout=32s 200 OK in 56 milliseconds
I0111 16:40:13.756919   45218 round_trippers.go:443] GET https://35.202.162.251/api/v1/namespaces/default/configmaps/test3 404 Not Found in 94 milliseconds
I0111 16:40:13.814068   45218 round_trippers.go:443] POST https://35.202.162.251/api/v1/namespaces/default/configmaps?dryRun=All&fieldManager=kubectl-client-side-apply 201 Created in 56 milliseconds
<diff snipped>
```

There are four calls to /openapi/v2, please remove 3.

I strongly suspect that most are getting the definitions from cache, but that's still one round-trip to the server, and I also strongly suspect that we re-decode the entire blob, which may take some time if you have A LOT of objects.

**Comment by apelisse** [comment]
I can't remember why the DryRun verifier is lazy?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/resource/dry_run_verifier.go#L81

**Comment by apelisse** [comment]
We're probably supposed to use this, to prevent this from happening again: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/util/openapi/openapi_getter.go#L43-L49

**Comment by julianvmodesto** [comment]
Oh, the dry run verifier is in k8s.io/cli-runtime, which should not depend on k8s.io/kubectl.

Also, the cached open API getter is an improvement, but we probably could change the dry run verifier to accept the open api resources object instead of any discovery client at all.

**Comment by apelisse** [comment]
> the dry run verifier to accept the OpenAPI schema object instead of any discovery client at all.

I'm sure there is a reason why it's lazy?

We could change the openapi getter to return a `func() Resources` so that we can get the resources lazily.

**Comment by apelisse** [comment]
On the other hand, we have downloaded it before we create the dryrun verifier here (in diff). I'm curious if there is a time where we need it lazy (I'm not convinced).

**Comment by julianvmodesto** [comment]
I don't think it needs to be lazy. For some reason the method signature changed in https://github.com/kubernetes/kubernetes/pull/86408 (see below), but I couldn't find any convincing reasoning. I think we can change it back to accept the schema instead of being lazy:

https://github.com/kubernetes/kubernetes/blob/0040920045e44692adfcefc85f959e680151cea2/staging/src/k8s.io/kubectl/pkg/util/openapi/dryrun.go#L50

**Comment by apelisse** [comment]
Ah, I see what's going on here. We typically hide the `openapi_v2.Document` behind the parsed form `Resources`. So we probably didn't have a good way to get the `Document`. We should probably split the thing that gets the `Document` from the thing that converts into `Resources`.

**Comment by apelisse** [comment]
Let's look for other places where we're using the `DryRunVerifier` (I'm sure we do that in apply too), because I don't see how it would avoid the bug. IOW, I think "apply" is also impacted.

**Comment by julianvmodesto** [comment]
Ah so we want the cached OpenAPI getter that gets a `Document` instead of the `Resources`. That could go into cli-runtime, and the existing cached getter can wrap it to return a `Resources`.

```
type Getter interface {
	Get() (Document, error)
}
```

Then the `DryRunVerifier` would accept one of these `Getter`s.

We use the `DryRunVerifier` in all commands and subcommands where we added `--dry-run=server`. I haven't come across any other internal/external uses.

Also, there's an existing cached getter in cmdutil:
https://github.com/kubernetes/kubernetes/blob/3a3445d926d4c7c7dac31ca959dd44cb60b01cc4/staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go#L163

**Comment by apelisse** [comment]
Yep, to be honest, I'd remove these `Getter` interfaces and replace them with mere functions, that would prevent us from having to propagate the dependencies all over.

**Comment by maknihamdi** [comment]
I think it's not only a diff issue. With my team we use kubectl with an all-in-one big yaml file to do our gitops checks and deployements, we see that we call openApi/v2 route hundreds times with every diff or apply command, you can imagine latency.
I thought about several solutions. I started with caching openApi document in diff command, and calling a special dryrunverifier.HasSupport function whitch can accept an openApi document (It is our quick win solution used with my team). We can do the same technique with other command supporting dryrun, caching openapi document and use it with a special HasSupport method, but we can do the same think with other global solutions

- have a new dryrunverifier implementation supporting cached openapi document ```NewDryRunVerifier(dynamicClient dynamic.Interface, openApi *openapi_v2.Document) *DryRunVerifier```

- keep the dryrunverifier implem and use a new discovery.DiscoveryInterface supporting memory cache, with a new implem of this func  ```func (d *CachedDiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {```

- Stop using DiscoveryInterface in DryRunVerifier and use an OpenApiGetter impl doing only once the openApi/v2 call (this is @julianvmodesto solution with this PR https://github.com/kubernetes/kubernetes/pull/98194)

So the question is: when do we use the cache? in the command (diff, apply, patch, ...) ? in DryRunVerifier? In discoveryinterface?

**Comment by julianvmodesto** [comment]
Fixed in 1.21 by #98194.

Keeping open since we're considering backporting a fix for past releases.
